### PR TITLE
[PS5][Driver] Pass default -z options to PS5 linker

### PR DESCRIPTION
**Pull request by playstation-edd**
Until now, these options have been hardcoded as downstream patches in LLD. Add them to the driver so that the private patches can be removed.

PS5 only. These implementation of these behaviours will remain in the proprietary linker on PS4.

SIE tracker: TOOLCHAIN-16704

### CODE REVIEW

**Comment by playstation-edd** [comment]
@pogo59 - I recall you expressed an interest in being subscribed to such PRs. Please do let me know if it gets tiresome!

**Comment by playstation-edd** [comment]
Note to self: fix typo in comment on submission ("The~se~ implementation ...").

**Review by jmorse — APPROVED** [review]
LGTM

File: clang/lib/Driver/ToolChains/PS4CPU.cpp [hunk 1]
```diff
@@ -240,11 +242,32 @@ void tools::PS5cpu::Linker::ConstructJob(Compilation &C, const JobAction &JA,
...
-  // Default to PIE for non-static executables.
-  const bool PIE =
-      !Args.hasArg(options::OPT_r, options::OPT_shared, options::OPT_static);
-  if (Args.hasFlag(options::OPT_pie, options::OPT_no_pie, PIE))
-    CmdArgs.push_back("-pie");
+  if (!Relocatable) {
+    // Default to PIE for non-static executables.
+    const bool PIE = !Args.hasArg(options::OPT_shared, options::OPT_static);
+    if (Args.hasFlag(options::OPT_pie, options::OPT_no_pie, PIE))
+      CmdArgs.push_back("-pie");
```
Comment (pogo59) [inline clang/lib/Driver/ToolChains/PS4CPU.cpp:249]: Did you mean to include these changes in the PR? It looks like the old code would pass through an explicit `-pie` even with `-r`; the new code will (probably) diagnose that `-pie` as unused. This is likely a good thing, but is not mentioned in the commit message.

Comment (playstation-edd) [reply clang/lib/Driver/ToolChains/PS4CPU.cpp:249]: The inclusion was deliberate. I took the opportunity to perform the check for `OPT_r` in one place. But I didn't spot the edge-case change in behaviour that you mention, forgetting that `hasArg()` had a side-effect - sorry!

Would we prefer that undo this bit and defer to a separate PR? Or update the final commit message and perhaps add a test case for the behaviour change?

Comment (playstation-edd) [reply clang/lib/Driver/ToolChains/PS4CPU.cpp]: I have undone this part of the change. Thanks for spotting it.

**Pull request merged**